nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,536 @@
+  layer->input_layers.push_back(current.getName());
+
+  if (current.num_outputs != 1)
+    return ML_ERROR_INVALID_PARAMETER;
+
+  layer->num_outputs = current.num_outputs;
+  layer->output_dim.resize(layer->num_outputs);
+  layer->output_layers.clear();
+  for (unsigned int i = 0; i < current.num_outputs; ++i)
+    layer->output_layers.push_back(current.output_layers[i]);
+
+  current.num_outputs = 1;
+  current.output_layers.clear();
+  current.output_layers.push_back(layer->getName());
+
+  addLayerNode(layer);
+
+  updateNameInLayers(layers, current.getName(), layer->getName());
+
+  return status;
+}
+
+int NetworkGraph::addLossLayer(const LossType loss_type) {
+  int status = ML_ERROR_NONE;
+  LossType updated_loss_type = loss_type;
+  if (num_node == 0) {
+    status = ML_ERROR_INVALID_PARAMETER;
+    NN_RETURN_STATUS();
+  }
+
+  if (updated_loss_type == LossType::LOSS_ENTROPY) {
+    if (getLayerNode(num_node - 1).layer->getType() != "activation") {
+      ml_loge("Error: Cross Entropy need last layer to have softmax or sigmoid "
+              "activation.");
+      return ML_ERROR_NOT_SUPPORTED;
+    }
+
+    LayerNode act_layer_node = getLayerNode(num_node - 1);
+    adj.pop_back();
+    num_node--;
+
+    switch (act_layer_node.layer->getActivationType()) {
+    case ActivationType::ACT_SIGMOID:
+      updated_loss_type = LossType::LOSS_ENTROPY_SIGMOID;
+      break;
+    case ActivationType::ACT_SOFTMAX:
+      updated_loss_type = LossType::LOSS_ENTROPY_SOFTMAX;
+      break;
+    default:
+      ml_loge("Error: Cross Entropy not supported without softmax or sigmoid.");
+      return ML_ERROR_NOT_SUPPORTED;
+    }
+  }
+
+  std::string input_str = getLayerNode(num_node - 1).layer->getName();
+
+  std::shared_ptr<LossLayer> layer = std::make_unique<LossLayer>();
+
+  ensureName(layer);
+
+  LayerNode last_node = getLayerNode(num_node - 1);
+  last_node.layer->num_outputs = 1;
+  last_node.layer->output_layers.clear();
+  last_node.layer->output_layers.push_back(layer->getName());
+
+  layer->num_inputs = 1;
+  layer->input_layers.clear();
+  layer->input_layers.push_back(input_str);
+
+  if (layer->output_layers.size() == 0) {
+    layer->num_outputs = 1;
+    layer->output_dim.resize(1);
+    layer->output_layers.push_back("__exit__");
+  }
+
+  status = layer->setLoss(updated_loss_type);
+  NN_RETURN_STATUS();
+
+  addLayerNode(layer);
+
+  return ML_ERROR_NONE;
+}
+
+void NetworkGraph::setOutputLayers(
+  std::vector<std::shared_ptr<Layer>> &layers) {
+
+  for (unsigned int idx = 0; idx < layers.size(); ++idx) {
+    unsigned int count = 0;
+    for (unsigned int i = 0; i < layers.size(); ++i) {
+      if (istrequal(layers[i]->getName(), layers[idx]->getName()))
+        continue;
+      for (unsigned int j = 0; j < layers[i]->input_layers.size(); ++j) {
+        if (istrequal(layers[i]->input_layers[j], layers[idx]->getName())) {
+          layers[idx]->output_layers.push_back(layers[i]->getName());
+          count++;
```

Comment:
How about using `layers[idx]->output_layers.size()` and remove count to reduce complexity?

Reply:
fixed.